File: app/utils/board_utils.py
import json

from lib import ndb_json
from models import Game


FLIPPED_POSITIONS = [9, 8, 7, 6, 5, 4, 3, 2, 1, 0]


def reverse_position(xy_pos):
    xy_pos['x'] = FLIPPED_POSITIONS[xy_pos['x']]
    xy_pos['y'] = FLIPPED_POSITIONS[xy_pos['y']]


def get_sendable_game(game, side):
    game_dict = json.loads(ndb_json.dumps(game))

    if side == 0:
        game_dict['player_hash'] = game_dict['red_hash']

    elif side == 1:
        game_dict['player_hash'] = game_dict['blue_hash']

    # These are secret and should never be sent.
    del game_dict['red_hash']
    del game_dict['blue_hash']
    del game_dict['red_setup']
    del game_dict['blue_setup']

    game_dict['side'] = side

    if game_dict['last_move']:
        game_dict['last_move'] = json.loads(game_dict['last_move'])

        # We need to reverse if we're blue
        if side == 1:
            reverse_position(game_dict['last_move']['to']['position'])
            reverse_position(game_dict['last_move']['from']['position'])

    else:
        del game_dict['last_move']

    game_dict['board'] = get_sendable_board(game, side)


    return game_dict


def get_sendable_board(game, side):
    board = game.get_board()

    # Only continue if the game hasn't finished
    if game.has_ended():
        if side == 1:
            return Game.reverse_board(board)
        else:
            return board

    if side == 0 and not game.blue_setup:
        board[0] = unknown_row(side)
        board[1] = unknown_row(side)
        board[2] = unknown_row(side)
        board[3] = unknown_row(side)

        return board

    else:
        hidden_side_board = hide_side(board, side)

        if side == 1:
            return Game.reverse_board(hidden_side_board)
        else:
            return hidden_side_board


def hide_side(board, side):
    for y in xrange(10):
        for x in xrange(10):
            if is_cell_occupied(board[y][x]) and board[y][x]['side'] != side:
                board[y][x] = unknown(side)

    return board


def is_cell_occupied(cell):
    if cell == 0 or cell == 1:
        return False
    else:
        return True


def unknown(side):
    return {'rank': 'U', 'side': OPPOSITE_SIDE[side]}


def unknown_row(side):
    return [unknown(side), unknown(side), unknown(side), unknown(side),
            unknown(side), unknown(side), unknown(side), unknown(side), unknown(side), unknown(side)]


OPPOSITE_SIDE = [1, 0]

File: src/board_utils.rs
use crate::models::{Board, Cell, Game, Piece, Position};

const FLIPPED: [usize; 10] = [9, 8, 7, 6, 5, 4, 3, 2, 1, 0];

/// Reverse a position for blue perspective: (x,y) → (9-x, 9-y)
pub fn reverse_position(pos: &mut Position) {
    pos.x = FLIPPED[pos.x];
    pos.y = FLIPPED[pos.y];
}

/// Build the game dict to send to a player, hiding sensitive data.
/// Port of board_utils.py get_sendable_game.
pub fn get_sendable_game(game: &Game, side: i32) -> serde_json::Value {
    let mut dict = serde_json::json!({
        "board": get_sendable_board(game, side),
        "turn": game.turn,
        "game_state": game.game_state,
        "side": side,
        "created": game.created,
        "modified": game.modified,
    });

    // Set player_hash based on side
    if side == 0 {
        dict["player_hash"] = serde_json::json!(game.red_hash);
    } else {
        dict["player_hash"] = serde_json::json!(game.blue_hash);
    }

    // Include join_hash for the creator (red) when waiting for opponent
    if side == 0 && game.game_state == 0 {
        dict["join_hash"] = serde_json::json!(game.join_hash);
    }

    // Set last_move (with position reversal for blue)
    if !game.moves.is_empty() {
        let mut last_move = game.get_last_move().unwrap();

        if side == 1 {
            // Reverse positions for blue
            if let Some(to) = last_move.get_mut("to") {
                if let Some(pos) = to.get_mut("position") {
                    let x = pos["x"].as_u64().unwrap() as usize;
                    let y = pos["y"].as_u64().unwrap() as usize;
                    pos["x"] = serde_json::json!(FLIPPED[x]);
                    pos["y"] = serde_json::json!(FLIPPED[y]);
                }
            }
            if let Some(from) = last_move.get_mut("from") {
                if let Some(pos) = from.get_mut("position") {
                    let x = pos["x"].as_u64().unwrap() as usize;
                    let y = pos["y"].as_u64().unwrap() as usize;
                    pos["x"] = serde_json::json!(FLIPPED[x]);
                    pos["y"] = serde_json::json!(FLIPPED[y]);
                }
            }
        }

        dict["last_move"] = last_move;
    }

    dict
}

/// Build the board to send to a player, hiding opponent pieces.
/// Port of board_utils.py get_sendable_board.
pub fn get_sendable_board(game: &Game, side: i32) -> Board {
    let mut board = game.board.clone();

    // If game has ended, show full board (rotated for blue)
    if game.has_ended() {
        if side == 1 {
            return Game::reverse_board(&board);
        }
        return board;
    }

    // Red waiting for blue: hide top rows as unknown
    if side == 0 && game.blue_setup.is_none() {
        let opposite = opposite_side(side);
        board[0] = unknown_row(opposite);
        board[1] = unknown_row(opposite);
        board[2] = unknown_row(opposite);
        board[3] = unknown_row(opposite);
        return board;
    }

    // Normal: hide opponent pieces
    hide_side(&mut board, side);

    if side == 1 {
        Game::reverse_board(&board)
    } else {
        board
    }
}

/// Replace opponent pieces with unknown markers.
pub fn hide_side(board: &mut Board, side: i32) {
    let opp = opposite_side(side);
    for row in board.iter_mut() {
        for cell in row.iter_mut() {
            if let Cell::Piece(p) = cell {
                if p.side != side {
                    *cell = Cell::Piece(Piece::unknown(opp));
                }
            }
        }
    }
}

fn opposite_side(side: i32) -> i32 {
    if side == 0 {
        1
    } else {
        0
    }
}

fn unknown_row(opponent_side: i32) -> Vec<Cell> {
    vec![Cell::Piece(Piece::unknown(opponent_side)); 10]
}
